mso/resource_mso_schema_template_contract_service_graph.go

```diff
@@ -1085,229 +392,83 @@ func resourceTemplateContractServiceGraphRead(d *schema.ResourceData, m interfac
-	if err != nil {
-		return nil, -1, fmt.Errorf("No Template found")
+	if graphSchema, ok := d.GetOk("service_graph_schema_id"); ok {
+		serviceGraphRef["schemaId"] = graphSchema.(string)
+	} else {
+		serviceGraphRef["schemaId"] = schemaId
 	}
 
-	for i := 0; i < templateCount; i++ {
-		templateCont, err := cont.ArrayElement(i, "templates")
-		if err != nil {
-			return nil, -1, fmt.Errorf("Unable to get template element")
-		}
-
-		apiTemplate := models.StripQuotes(templateCont.S("name").String())
-
-		if apiTemplate == templateName {
-			log.Printf("[DEBUG] Template found")
-
-			sgCount, err := templateCont.ArrayCount("serviceGraphs")
-
-			if err != nil {
-				return nil, -1, fmt.Errorf("No Service Graph found")
-			}
-
-			for j := 0; j < sgCount; j++ {
-				sgCont, err := templateCont.ArrayElement(j, "serviceGraphs")
-
-				if err != nil {
-					return nil, -1, fmt.Errorf("Unable to get service graph element")
-				}
-
-				apiSgName := models.StripQuotes(sgCont.S("name").String())
-
-				if apiSgName == graphName {
-					return sgCont, j, nil
-				}
-			}
-
-		}
+	if graphTemp, ok := d.GetOk("service_graph_template_name"); ok {
+		serviceGraphRef["templateName"] = graphTemp.(string)
+	} else {
+		serviceGraphRef["templateName"] = TemplateName
 	}
 
-	return nil, -1, fmt.Errorf("unable to find service graph")
-}
-
-func extractNodes(cont *container.Container) ([]interface{}, error) {
-	nodes := make([]interface{}, 0, 1)
-	count, err := cont.ArrayCount("serviceNodes")
+	cont, err := msoClient.GetViaURL(fmt.Sprintf("api/v1/schemas/%s", serviceGraphRef["schemaId"].(string)))
 	if err != nil {
-		return nodes, err
+		return err
 	}
 
-	for i := 0; i < count; i++ {
-		node, err := cont.ArrayElement(i, "serviceNodes", "name")
-		if err != nil {
-			return nodes, err
-		}
+	graphCont, _, err := getSchemaTemplateServiceGraph(cont, serviceGraphRef["templateName"].(string), serviceGraph)
 
-		nodes = append(nodes, models.StripQuotes(node.String()))
+	if err != nil {
+		return err
 	}
-	return nodes, nil
-}
-
-func getSiteServiceNode(graphCont *container.Container, schemaId, templateName, graphName, nodeName string) (*container.Container, int, error) {
-
-	nodesCount, err := graphCont.ArrayCount("serviceNodes")
+	nodeList, err := extractNodes(graphCont)
 	if err != nil {
-		return nil, -1, fmt.Errorf("Unable to load count site service node")
+		return err
 	}
-	for i := 0; i < nodesCount; i++ {
-		nodeCont, err := graphCont.ArrayElement(i, "serviceNodes")
-		if err != nil {
-			return nil, -1, fmt.Errorf("Unable to site service node element")
-		}
 
-		nodeRef := models.StripQuotes(nodeCont.S("serviceNodeRef").String())
-
-		nodeSplit := strings.Split(nodeRef, "/")
-		if len(nodeSplit) == 9 {
-			if nodeSplit[2] == schemaId && nodeSplit[4] == templateName && nodeSplit[6] == graphName && nodeSplit[8] == nodeName {
-				return nodeCont, i, nil
+	tempNodeList := d.Get("node_relationship").([]interface{})
```

Comment:
should we rename "tempNodeList" and "nodeList" to better distinguish difference between from schema input and from ndo node configuration?